src/server/api.rs

```diff
@@ -0,0 +1,254 @@
+            "activation_source": activation_source,
+        });
+
+        let mut response = client
+            .patch(format!("{}/users/{}/activate", self.api_url, user_id))
+            .header("X-Api-Key", self.api_key.as_str())
+            .send_json(&json)
+            .await
+            .map_err(|e| ApiError::SendRequestError(e.to_string()))?;
+
+        let user: DisplayUser = match response.status() {
+            StatusCode::OK => response.json().await?,
+            status => return Err(ApiError::HttpError(status.to_string())),
+        };
+
+        Ok(user)
+    }
+
+    /// Log in user via alternate_id and password
+    pub async fn user_login(
+        &self,
+        alternate_id: String,
+        password: String,
+    ) -> Result<Auth, ApiError> {
+        let client = Client::default();
+        let json = json!({
+            "alternate_id": alternate_id,
+            "password": password,
+        });
+
+        let mut response = client
+            .post(format!("{}/auth/token", self.api_url))
+            .header("X-Api-Key", self.api_key.as_str())
+            .send_json(&json)
+            .await
+            .map_err(|e| ApiError::SendRequestError(e.to_string()))?;
+
+        // todo: error messages should be improved
+        let auth: Auth = match response.status() {
+            StatusCode::OK => response.json().await?,
+            status => return Err(ApiError::HttpError(status.to_string())),
+        };
+
+        Ok(auth)
+    }
+
+    /// Request a random Yat for the user
+    pub async fn random_yat(
+        &self,
+        access_token: String,
+        code_id: Uuid,
+        signature: Signature,
+        pubkey: RistrettoPublicKey,
+    ) -> Result<DisplayOrder, ApiError> {
+        let client = Client::default();
+        let json = json!({
+            "nonce": signature.get_public_nonce().to_hex(),
+            "signature": signature.get_signature().to_hex(),
+            "pubkey": pubkey.to_hex(),
+            "tracking_data": {"source": "Yat Partner API"},
+        });
+
+        let mut response = client
+            .post(format!("{}/codes/{}/random_yat", self.api_url, code_id))
+            .bearer_auth(access_token)
+            .send_json(&json)
+            .await
+            .map_err(|e| ApiError::SendRequestError(e.to_string()))?;
+
+        // todo: error messages should be improved
+        let order: DisplayOrder = match response.status() {
+            StatusCode::OK => response.json().await?,
+            status => return Err(ApiError::HttpError(status.to_string())),
+        };
+
+        Ok(order)
+    }
+
+    /// Checkout the cart contents for the user
+    pub async fn checkout(
+        &self,
+        access_token: String,
+        pubkey: RistrettoPublicKey,
+    ) -> Result<DisplayOrder, ApiError> {
+        let client = Client::default();
+        let json = json!({
+            "method": "Free",
+            "payment_method_id": "ec18d2d7-d2e0-41e4-98e4-847f14422d8a",
```

Comment:
Where does this UUID come from?

Reply:
From the [docs](http://yat-api-doc-preview.s3-website-us-east-1.amazonaws.com/docs/api-ref/#checkout-cart-with-provided-payment-details)

Reply:
which actually need an update too, I had to change "method" from "Provider" to "Free"

Reply:
Yeah the docs seem out of date. That object has changed.
```rust
pub struct CheckoutCartRequest {
    /// Payment method type
    pub method: PaymentProviders,
    /// URL user will be redirected after successful payment
    /// Required for Stripe Checkout
    pub success_url: Option<String>,
    /// URL user will be redirected if payment cancelled
    /// Required for Stripe Checkout
    pub cancel_url: Option<String>,
    /// Optional: tracking data
    pub tracking_data: Option<serde_json::Value>,
    /// Optional: The user's public key to associate with this emoji id. If provided will use this pubkey otherwise
    /// will default to the first pubkey returned for the user.
    pub pubkey: Option<Pubkey>,
    /// Amount paid in cash. Applicable and required only for `Cash` payment option for Admin.
    pub amount: Option<i64>,
    /// External reference for cash payment. Applicable and required only for `Cash` payment option for Admin.
    pub external_reference: Option<String>,
}
```

Reply:
Nice one yeah I saw that. How should the payment_method_id be populated?

Reply:
You can drop it entirely.

Reply:
Nice have done so, and also dropped other unnecessary keys